File: dotnet/RailsLib.Net/Net/Game/TileUpgrade.cs
using GameLib.Net.Common;
using GameLib.Net.Common.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GameLib.Net.Game
{
    public class TileUpgrade : IUpgrade
    {
        private static Logger<TileUpgrade> log = new Logger<TileUpgrade>();

        /**
         * Rotation defines the following details for a tile upgrade
         */
        public class Rotation
        {
            private HexSidesSet connectedSides;
            private HexSidesSet sidesWithNewTrack;
            private HexSide rotation;
            private Dictionary<Station, Station> stationMapping;
            private List<Station> stationsWithNewTrack;
            private bool symmetric;

            public Rotation(List<Track> connectedTracks, List<Track> newTracks, HexSide rotation,
                    Dictionary<Station, Station> mapping, bool symmetric)
            {

                this.rotation = rotation;
                this.stationMapping = mapping;
                this.symmetric = symmetric;

                HexSidesSet.Builder sidesBuilder = HexSidesSet.GetBuilder();
                foreach (Track t in connectedTracks)
                {
                    if (t.Start.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE)
                    {
                        sidesBuilder.Set((HexSide)t.Start);
                    }
                    if (t.End.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE)
                    {
                        sidesBuilder.Set((HexSide)t.End);
                    }
                }
                connectedSides = sidesBuilder.Build();

                sidesBuilder = HexSidesSet.GetBuilder();
                List<Station> stationBuilder = new List<Station>();
                foreach (Track t in newTracks)
                {
                    if (t.Start.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE)
                    {
                        sidesBuilder.Set((HexSide)t.Start);
                    }
                    else
                    {
                        stationBuilder.Add((Station)t.Start);
                    }
                    if (t.End.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE)
                    {
                        sidesBuilder.Set((HexSide)t.End);
                    }
                    else
                    {
                        stationBuilder.Add((Station)t.End);
                    }
                }
                stationsWithNewTrack = stationBuilder;

                // Special condition for restrictive tile lays:
                // If a station with new track has more slots then the replaced station of the base tile
                // then all sides connecting to the station of the base tile are considered as
                // sides with new track as well
                if (stationMapping != null && stationsWithNewTrack.Count > 0)
                {
                    foreach (Track t in connectedTracks)
                    {
                        if (t.Start.TrackPointType == TrackPoint.TrackPointTypeEnum.STATION
                                && t.End.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE)
                        {
                            Station start = (Station)t.Start;
                            if (stationsWithNewTrack.Contains(start)
                                    && stationMapping.ContainsKey(start)
                                    && start.BaseSlots
                                    < stationMapping[start].BaseSlots)
                            {
                                sidesBuilder.Set((HexSide)t.End);
                            }
                        }
                    }
                }
                sidesWithNewTrack = sidesBuilder.Build();
            }

            public HexSidesSet ConnectedSides
            {
                get
                {
                    return connectedSides;
                }
            }

            public HexSidesSet SidesWithNewTrack
            {
                get
                {
                    return sidesWithNewTrack;
                }
            }

            public Dictionary<Station, Station> StationMapping
            {
                get
                {
                    return stationMapping;
                }
            }

            public List<Station> StationsWithNewTrack
            {
                get
                {
                    return stationsWithNewTrack;
                }
            }

            public bool IsSymmetric
            {
                get
                {
                    return symmetric;
                }
            }

            override public string ToString()
            {
                return "rotation = " + rotation.ToString() + ", connectedSides = "
                        + connectedSides.ToString() + ", sidesWithNewTrack = " + sidesWithNewTrack.ToString()
                        + ", stationMapping = " + stationMapping + ", stationsWithNewTrack = " + stationsWithNewTrack;
            }
        }

        /** Tile to upgrade */
        private Tile baseTile;

        /** The upgrade tile number */
        private string targetTileId;

        /**
         * Temporary Strings to exclude hexes and phases. This will be
         * processed at finishConfiguration.
         */
        private string hexes;
        private string phases;

        /** The upgrade tile */
        private Tile targetTile;

        /** Possible rotations given the trackConfiguration */
        private Dictionary<HexSide, Rotation> rotations;
        private HexSidesSet rotationSides;

        /** Hexes where the upgrade can be executed */
        private List<MapHex> allowedHexes = null;

        /**
         * Hexes where the upgrade cannot be executed Only one of allowedHexes
         * and disallowedHexes should be used
         */
        private List<MapHex> disallowedHexes = null;

        /**
         * Phases in which the upgrade can be executed.
         */
        private List<Phase> allowedPhases = null;


        private TileUpgrade(Tile baseTile, string targetId, string hexes, string phases)
        {
            this.baseTile = baseTile;
            this.targetTileId = targetId;
            this.hexes = hexes;
            this.phases = phases;
        }

        public static List<TileUpgrade> CreateFromTags(Tile tile, List<Tag> upgradeTags)
        {

            List<TileUpgrade> allUpgrades = new List<TileUpgrade>();
            foreach (Tag upgradeTag in upgradeTags)
            {
                string ids = upgradeTag.GetAttributeAsString("id");
                if (ids == null) continue;

                string hexes = upgradeTag.GetAttributeAsString("hex");
                string phases = upgradeTag.GetAttributeAsString("phase");

                foreach (string sid in ids.Split(','))
                {
                    try
                    {
                        TileUpgrade upgrade = new TileUpgrade(tile, sid, hexes, phases);
                        allUpgrades.Add(upgrade);
                    }
                    catch (FormatException e)
                    {
                        log.Error("Caught Exception : " + e.Message);
                        throw new ConfigurationException(LocalText.GetText(
                                "InvalidUpgrade", tile.ToText(), sid));
                    }
                }
            }
            return allUpgrades;
        }

        public static TileUpgrade CreateSpecific(Tile current, Tile specific)
        {
            TileUpgrade upgrade = new TileUpgrade(current, specific.Id, null, null);
            try
            {
                upgrade.FinishConfiguration(current.GetRoot);
            }
            catch (ConfigurationException)
            {
                log.Error(LocalText.GetText("InvalidUpgrade",
                        current.ToText(), specific.ToText()));
            }
            return upgrade;
        }

        public void FinishConfiguration(RailsRoot root)
        {
            targetTile = root.TileManager.GetTile(targetTileId);
            if (targetTile == null)
            {
                throw new ConfigurationException(LocalText.GetText("InvalidUpgrade",
                        baseTile.ToText(), targetTileId));
            }
            InitRotations();
            ParsePhases(root);
            ParseHexes(root);
        }

        public Tile TargetTile
        {
            get
            {
                return targetTile;
            }
        }

        public string TileId
        {
            get
            {
                return targetTileId;
            }
        }

        public bool IsAllowedForHex(MapHex hex)
        {
            if (allowedHexes != null)
            {
                return allowedHexes.Contains(hex);
            }
            else if (disallowedHexes != null)
            {
                return !disallowedHexes.Contains(hex);
            }
            else
            {
                return true;
            }
        }

        public bool IsAllowedForPhase(Phase phase)
        {
            if (allowedPhases != null
                    && !allowedPhases.Contains(phase))
            {
                return false;
            }
            else
            {
                return true;
            }
        }

        private void InitRotations()
        {
            HexSidesSet.Builder sideBuilder = HexSidesSet.GetBuilder();
            Dictionary<HexSide, Rotation> rotationBuilder = new Dictionary<HexSide, Rotation>();
            foreach (HexSide side in HexSide.All())
            {
                Rotation rotation = ProcessRotations(side);
                if (rotation != null)
                {
                    sideBuilder.Set(side);
                    rotationBuilder[side] = rotation;
                }
            }
            rotationSides = sideBuilder.Build();
            rotations = rotationBuilder;
        }

        private void ParsePhases(RailsRoot root)
        {
            if (phases == null) return;

            List<Phase> phaseBuilder = new List<Phase>();
            foreach (string phaseName in phases.Split(','))
            {
                Phase phase = root.PhaseManager.GetPhaseByName(phaseName);
                if (phase == null)
                {
                    throw new ConfigurationException(LocalText.GetText("IllegalPhaseDefinition", this.ToString()));
                }
                else
                {
                    phaseBuilder.Add(phase);
                }
            }
            allowedPhases = phaseBuilder;
        }

        private List<MapHex> ParseHexString(RailsRoot root, string sHexes)
        {
            List<MapHex> hexBuilder = new List<MapHex>();

            foreach (string sHex in sHexes.Split(','))
            {
                MapHex hex = root.MapManager.GetHex(sHex);
                if (hex == null)
                {
                    throw new ConfigurationException(LocalText.GetText("InvalidUpgrade", baseTile.ToText(), sHexes));
                }
                else
                {
                    hexBuilder.Add(hex);
                }
            }
            return hexBuilder;
        }

        private void ParseHexes(RailsRoot root)
        {
            if (hexes == null) return;

            bool allowed = !hexes.StartsWith("-");
            if (allowed)
            {
                allowedHexes = ParseHexString(root, hexes);
            }
            else
            {
                disallowedHexes = ParseHexString(root, hexes.Substring(1));
            }
        }


        private bool CheckInvalidSides(Rotation rotation, HexSidesSet impassable)
        {
            log.Debug("Check rotation " + rotation + " against  impassable" + impassable);
            if (impassable == null) return false; // null implies that no station exists
            return (rotation.ConnectedSides.Intersects(impassable));
        }

        private bool CheckSideConnectivity(Rotation rotation, HexSidesSet connected, bool restrictive)
        {
            log.Debug("Check rotation " + rotation + " against " + connected);
            if (connected == null) return true; // null implies no connectivity required
            if (restrictive && !rotation.SidesWithNewTrack.IsEmpty)
            {
                return rotation.SidesWithNewTrack.Intersects(connected);
            }
            else
            {
                return (rotation.ConnectedSides.Intersects(connected));
            }
        }

        private bool CheckStationConnectivity(Rotation rotation, IEnumerable<Station> stations)
        {
            if (rotation.StationMapping == null) return false;
            log.Debug("Check Stations " + stations + ", rotation = " + rotation);
            foreach (Station station in stations)
            {
                Station targetStation;
                rotation.StationMapping.TryGetValue(station, out targetStation);
                if (targetStation != null && rotation.StationsWithNewTrack.Contains(targetStation))
                {
                    return true;
                }
            }
            return false;
        }

        public HexSidesSet RotationSet
        {
            get
            {
                return rotationSides;
            }
        }

        public Rotation GetRotation(HexSide rotation)
        {
            return rotations[rotation];
        }

        public HexSidesSet GetAllowedRotations(HexSidesSet connected, HexSidesSet impassable, HexSide baseRotation,
                IEnumerable<Station> stations, bool restrictive)
        {

            HexSidesSet.Builder builder = HexSidesSet.GetBuilder();
            foreach (HexSide side in rotationSides)
            {
                Rotation rotation = rotations[side];
                if (CheckInvalidSides(rotation, impassable)) continue;
                if (CheckSideConnectivity(rotation, connected, restrictive) ||
                        CheckStationConnectivity(rotation, stations))
                {
                    builder.Set((HexSide)side.Rotate(baseRotation));
                }
            }
            HexSidesSet allowed = builder.Build();
            log.Debug("allowed = " + allowed + "hexSides = " + connected + "impassable =" + impassable + " rotationSides = " + rotationSides);
            return allowed;
        }

        private Rotation ProcessRotations(HexSide side)
        {

            TrackConfig baseConfig = baseTile.TrackConfig;
            TrackConfig target = targetTile.TrackConfig;
            // create rotation of target, unless default (= 0) rotation
            if (side != HexSide.Get(0))
            {
                target = TrackConfig.CreateByRotation(target, side);
            }
            // check if there are stations to map
            Dictionary<Station, Station> stationMapping = AssignStations(baseConfig, target);
            if (stationMapping != null && stationMapping.Count > 0)
            {
                if (stationMapping.ContainsValue(null))
                {
                    baseConfig = TrackConfig.CreateByDowngrade(baseConfig, baseConfig.Tile.GetStation(1));
                }
                baseConfig = TrackConfig.CreateByStationMapping(baseConfig, stationMapping);
            }

            // and finally check if all tracks are maintained
            List<Track> baseTracks = baseConfig.Tracks;
            List<Track> targetTracks = target.Tracks;

            List<Track> diffTrack = new List<Track>(baseTracks.Except(targetTracks));
            if (diffTrack.Count == 0)
            {
                IEnumerable<Track> newTracks = targetTracks.Except(baseTracks);
                bool allowed = (targetTile.PossibleRotations.Get(side));
                Rotation rotObject = new Rotation(targetTracks, new List<Track>(newTracks), side, stationMapping, allowed);
                log.Debug("New Rotation for " + baseTile + " => " + targetTile + ": \n" + rotObject);
                return rotObject;
            }
            else
            {
                log.Debug("No Rotation found " + baseTile + " => " + targetTile + ", rotation =" + side +
                        ", remaining Tracks = " + diffTrack);
                return null;
            }
        }

        private Dictionary<Station, Station> AssignStations(TrackConfig baseConfig, TrackConfig target)
        {
            int baseNb = baseConfig.Tile.NumStations;
            if (baseNb == 0) return null;

            int targetNb = target.Tile.NumStations;
            Dictionary<Station, Station> stationMap = new Dictionary<Station, Station>(baseNb);//Maps.newHashMapWithExpectedSize(baseNb);
            if (baseNb == 1)
            {
                // only one station in base => baseStation
                Station baseStation = baseConfig.Tile.GetStation(1);
                if (targetNb == 1)
                {
                    // only one station in target => targetStation
                    Station targetStation = target.Tile.GetStation(1);
                    // default case: 1 => 1 mapping
                    stationMap[baseStation] = targetStation;
                }
                else if (targetNb == 0)
                {
                    // special case: downgrade in 1856 and there is only one station to consider
                    /*List<TrackPoint>*/var baseTrack = baseConfig.GetStationTracks(baseStation);
                    foreach (TrackPoint side in baseTrack)
                    {
                        List<TrackPoint> targetTrack = target.GetSideTracks((HexSide)side);
                        targetTrack.Add(side); // connectivity with all other sides
                        List<TrackPoint> diffTrack = new List<TrackPoint>(baseTrack.Except(targetTrack));
                        if (diffTrack.Count > 0) return null;
                    }
                    stationMap[baseStation] = null;
                }
            }
            else
            { // more than one base station, assign by side connectivity
                List<Station> noTrackBaseStations = new List<Station>();
                SortedSet<Station> targetStations = new SortedSet<Station>(target.Tile.Stations);
                foreach (Station b in baseConfig.Tile.Stations)
                {
                    /*List<TrackPoint>*/var baseTrack = baseConfig.GetStationTracks(b);
                    if (baseTrack.Count == 0)
                    { // if track is empty, keep track to add target later
                        noTrackBaseStations.Add(b);
                    }
                    else
                    {
                        foreach (Station t in target.Tile.Stations)
                        {
                            /*List<TrackPoint>*/var targetTrack = target.GetStationTracks(t);
                            if (CheckTrackConnectivity(baseTrack, targetTrack))
                            {
                                stationMap[b] = t;
                                targetStations.Remove(t);
                                break;
                            }
                        }
                    }
                }
                // any base Stations remaining
                foreach (Station b in noTrackBaseStations)
                {
                    Station t = targetStations.Min;
                    if (t != null)
                    {
                        targetStations.Remove(t);
                        stationMap[b] = t;
                    }
                }
                // check if all base and target stations are assigned
                if (stationMap.Keys.Count != baseNb ||
                        stationMap.Values.Count != targetNb)
                {
                    stationMap = null;
                    log.Debug("Mapping: Not all stations assigned, set stationMap to null");
                }
            }
            return stationMap;
        }

        private bool CheckTrackConnectivity(IEnumerable<TrackPoint> baseTrack, IEnumerable<TrackPoint> targetTrack)
        {
            List<TrackPoint> diffTrack = new List<TrackPoint>(baseTrack.Except(targetTrack));
            if (diffTrack.Count == 0)
            {
                // target maintains connectivity
                return true;
            }
            else
            {
                //    // if not all connections are maintained, 
                //    Predicate<TrackPoint> checkForStation = new Predicate<TrackPoint>()
                //    {
                //            public bool apply(TrackPoint p)
                //    {
                //        return (p.getTrackPointType() == TrackPoint.Type.SIDE);
                //    }
                //};
                // check if remaining tracks only lead to other stations 
                //if (Sets.filter(diffTrack, checkForStation).isEmpty())
                //{
                //    return true;
                //}
                if (!diffTrack.Any(p => p.TrackPointType == TrackPoint.TrackPointTypeEnum.SIDE))
                {
                    return true;
                }
            }
            return false;
        }

        override public string ToString()
        {
            //return Objects.toStringHelper(this).add("base", baseTile).add("targetTile", targetTile).toString();
            return $"{this.GetType().Name}{{base={baseTile.ToString()}}}{{targetTile={targetTile.ToString()}}}";
        }
    }
}

File: src/main/java/net/sf/rails/game/TileUpgrade.java
package net.sf.rails.game;

import com.google.common.base.MoreObjects;
import com.google.common.base.Predicate;
import com.google.common.collect.*;
import com.google.common.collect.Sets.SetView;
import net.sf.rails.common.LocalText;
import net.sf.rails.common.parser.ConfigurationException;
import net.sf.rails.common.parser.Tag;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class TileUpgrade implements Upgrade {

    private static final Logger log = LoggerFactory.getLogger(TileUpgrade.class);

    /**
     * Rotation defines the following details for a tile upgrade
     */
    public static class Rotation {
        private final HexSidesSet connectedSides;
        private final HexSidesSet sidesWithNewTrack;
        private final HexSide rotation;
        private final Map<Station, Station> stationMapping;
        private final Set<Station> stationsWithNewTrack;
        private final boolean symmetric;

        private Rotation(Set<Track> connectedTracks, Set<Track> newTracks, HexSide rotation,
                         Map<Station, Station> mapping, boolean symmetric) {

            this.rotation = rotation;
            this.stationMapping = mapping;
            this.symmetric = symmetric;

            HexSidesSet.Builder sidesBuilder = HexSidesSet.builder();
            for (Track t : connectedTracks) {
                if (t.getStart().getTrackPointType() == TrackPoint.Type.SIDE) {
                    sidesBuilder.set((HexSide) t.getStart());
                }
                if (t.getEnd().getTrackPointType() == TrackPoint.Type.SIDE) {
                    sidesBuilder.set((HexSide) t.getEnd());
                }
            }
            connectedSides = sidesBuilder.build();

            sidesBuilder = HexSidesSet.builder();
            ImmutableSet.Builder<Station> stationBuilder = ImmutableSet.builder();
            for (Track t : newTracks) {
                if (t.getStart().getTrackPointType() == TrackPoint.Type.SIDE) {
                    sidesBuilder.set((HexSide) t.getStart());
                } else {
                    stationBuilder.add((Station) t.getStart());
                }
                if (t.getEnd().getTrackPointType() == TrackPoint.Type.SIDE) {
                    sidesBuilder.set((HexSide) t.getEnd());
                } else {
                    stationBuilder.add((Station) t.getEnd());
                }
            }
            stationsWithNewTrack = stationBuilder.build();

            // Special condition for restrictive tile lays:
            // If a station with new track has more slots then the replaced station of the base tile
            // then all sides connecting to the station of the base tile are considered as
            // sides with new track as well
            if (stationMapping != null && !stationsWithNewTrack.isEmpty()) {
                for (Track t : connectedTracks) {
                    if (t.getStart().getTrackPointType() == TrackPoint.Type.STATION
                            && t.getEnd().getTrackPointType() == TrackPoint.Type.SIDE) {
                        Station start = (Station) t.getStart();
                        if (stationsWithNewTrack.contains(start)
                                && stationMapping.containsKey(start)
                                && start.getBaseSlots()
                                < stationMapping.get(start).getBaseSlots()) {
                            sidesBuilder.set((HexSide) t.getEnd());
                        }
                    }
                }
            }
            sidesWithNewTrack = sidesBuilder.build();
        }

        public HexSidesSet getConnectedSides() {
            return connectedSides;
        }

        public HexSidesSet getSidesWithNewTrack() {
            return sidesWithNewTrack;
        }

        public Map<Station, Station> getStationMapping() {
            return stationMapping;
        }

        public Set<Station> getStationsWithNewTrack() {
            return stationsWithNewTrack;
        }

        public boolean isSymmetric() {
            return symmetric;
        }

        @Override
        public String toString() {
            return new StringBuilder().append("rotation = ").append(rotation).append(", connectedSides = ").
                    append(connectedSides.toString()).append(", sidesWithNewTrack = ").append(sidesWithNewTrack.toString()).
                    append(", stationMapping = ").append(stationMapping).append(", stationsWithNewTrack = ").append(stationsWithNewTrack).toString();
        }
    }

    /**
     * Tile to upgrade
     */
    private final Tile baseTile;

    /**
     * The upgrade tile number
     */
    private final String targetTileId;

    /**
     * Temporary Strings to exclude hexes and phases. This will be
     * processed at finishConfiguration.
     */
    private final String hexes;
    private final String phases;

    /**
     * The upgrade tile
     */
    private Tile targetTile;

    /**
     * Possible rotations given the trackConfiguration
     */
    private Map<HexSide, Rotation> rotations;
    private HexSidesSet rotationSides;

    /**
     * Hexes where the upgrade can be executed
     */
    private List<MapHex> allowedHexes = null;

    /**
     * Hexes where the upgrade cannot be executed Only one of allowedHexes
     * and disallowedHexes should be used
     */
    private List<MapHex> disallowedHexes = null;

    /**
     * Phases in which the upgrade can be executed.
     */
    private List<Phase> allowedPhases = null;


    private TileUpgrade(Tile baseTile, String targetId, String hexes, String phases) {
        this.baseTile = baseTile;
        this.targetTileId = targetId;
        this.hexes = hexes;
        this.phases = phases;
    }

    public static List<TileUpgrade> createFromTags(Tile tile, List<Tag> upgradeTags)
            throws ConfigurationException {

        ImmutableList.Builder<TileUpgrade> allUpgrades = ImmutableList.builder();
        for (Tag upgradeTag : upgradeTags) {
            String ids = upgradeTag.getAttributeAsString("id");
            if (ids == null) continue;

            String hexes = upgradeTag.getAttributeAsString("hex");
            String phases = upgradeTag.getAttributeAsString("phase");

            for (String sid : ids.split(",")) {
                try {
                    TileUpgrade upgrade = new TileUpgrade(tile, sid, hexes, phases);
                    allUpgrades.add(upgrade);
                } catch (NumberFormatException e) {
                    log.error("Catched Exception", e);
                    throw new ConfigurationException(LocalText.getText(
                            "InvalidUpgrade", tile.toText(), sid));
                }
            }
        }
        return allUpgrades.build();
    }

    public static TileUpgrade createSpecific(Tile current, Tile specific) {
        TileUpgrade upgrade = new TileUpgrade(current, specific.getId(), null, null);
        try {
            upgrade.finishConfiguration(current.getRoot());
        } catch (ConfigurationException e) {
            log.error(LocalText.getText("InvalidUpgrade",
                    current.toText(), specific.toText()), e);
        }
        return upgrade;
    }

    public void finishConfiguration(RailsRoot root) throws ConfigurationException {
        targetTile = root.getTileManager().getTile(targetTileId);
        if (targetTile == null) {
            throw new ConfigurationException(LocalText.getText("InvalidUpgrade",
                    baseTile.toText(), targetTileId));
        }
        initRotations();
        parsePhases(root);
        parseHexes(root);
    }

    public Tile getTargetTile() {
        return targetTile;
    }

    public String getTileId() {
        return targetTileId;
    }

    public boolean isAllowedForHex(MapHex hex) {
        if (allowedHexes != null) {
            return allowedHexes.contains(hex);
        } else if (disallowedHexes != null) {
            return !disallowedHexes.contains(hex);
        } else {
            return true;
        }
    }

    public boolean isAllowedForPhase(Phase phase) {
        if (allowedPhases != null
                && !allowedPhases.contains(phase)) {
            return false;
        } else {
            return true;
        }
    }

    private void initRotations() {
        HexSidesSet.Builder sideBuilder = HexSidesSet.builder();
        ImmutableMap.Builder<HexSide, Rotation> rotationBuilder = ImmutableMap.builder();
        for (HexSide side : HexSide.all()) {
            Rotation rotation = processRotations(side);
            if (rotation != null) {
                sideBuilder.set(side);
                rotationBuilder.put(side, rotation);
            }
        }
        rotationSides = sideBuilder.build();
        rotations = rotationBuilder.build();
    }

    private void parsePhases(RailsRoot root) throws ConfigurationException {
        if (phases == null) return;

        ImmutableList.Builder<Phase> phaseBuilder = ImmutableList.builder();
        for (String phaseName : phases.split(",")) {
            Phase phase = root.getPhaseManager().getPhaseByName(phaseName);
            if (phase == null) {
                throw new ConfigurationException(LocalText.getText(
                        "IllegalPhaseDefinition",
                        this.toString()
                ));
            } else {
                phaseBuilder.add(phase);
            }
        }
        allowedPhases = phaseBuilder.build();
    }

    private List<MapHex> parseHexString(RailsRoot root, String sHexes)
            throws ConfigurationException {
        ImmutableList.Builder<MapHex> hexBuilder = ImmutableList.builder();

        for (String sHex : sHexes.split(",")) {
            MapHex hex = root.getMapManager().getHex(sHex);
            if (hex == null) {
                throw new ConfigurationException(LocalText.getText("InvalidUpgrade",
                        baseTile.toText(), sHexes));
            } else {
                hexBuilder.add(hex);
            }
        }
        return hexBuilder.build();
    }

    private void parseHexes(RailsRoot root) throws ConfigurationException {
        if (hexes == null) return;

        boolean allowed = !hexes.startsWith("-");
        if (allowed) {
            allowedHexes = parseHexString(root, hexes);
        } else {
            disallowedHexes = parseHexString(root, hexes.substring(1));
        }
    }


    private boolean checkInvalidSides(Rotation rotation, HexSidesSet impassable) {
        log.trace("Check rotation {} against  impassable{}", rotation, impassable);
        if (impassable == null) return false; // null implies that no station exists
        return (rotation.getConnectedSides().intersects(impassable));
    }

    private boolean checkSideConnectivity(Rotation rotation, HexSidesSet connected, boolean restrictive) {
        log.trace("Check rotation {} against {}", rotation, connected);
        if (connected == null) return true; // null implies no connectivity required
        if (restrictive && !rotation.getSidesWithNewTrack().isEmpty()) {
            return rotation.getSidesWithNewTrack().intersects(connected);
        } else {
            return (rotation.getConnectedSides().intersects(connected));
        }
    }

    private boolean checkStationConnectivity(Rotation rotation, Collection<Station> stations) {
        if (rotation.getStationMapping() == null) return false;
        log.trace("Check Stations {}, rotation = {}", stations, rotation);
        for (Station station : stations) {
            Station targetStation = rotation.getStationMapping().get(station);
            if (targetStation != null && rotation.getStationsWithNewTrack().contains(targetStation)) {
                return true;
            }
        }
        return false;
    }

    public HexSidesSet getRotationSet() {
        return rotationSides;
    }

    public Rotation getRotation(HexSide rotation) {
        return rotations.get(rotation);
    }

    public HexSidesSet getAllowedRotations(HexSidesSet connected, HexSidesSet impassable, HexSide baseRotation,
                                           Collection<Station> stations, boolean restrictive) {

        HexSidesSet.Builder builder = HexSidesSet.builder();
        for (HexSide side : rotationSides) {
            Rotation rotation = rotations.get(side);
            if (checkInvalidSides(rotation, impassable)) continue;
            if (checkSideConnectivity(rotation, connected, restrictive) ||
                    checkStationConnectivity(rotation, stations)) {
                builder.set(side.rotate(baseRotation));
            }
        }
        HexSidesSet allowed = builder.build();
        log.trace("allowed = {}hexSides = {}impassable ={} rotationSides = {}", allowed, connected, impassable, rotationSides);
        return allowed;
    }

    private Rotation processRotations(HexSide side) {

        TrackConfig base = baseTile.getTrackConfig();
        TrackConfig target = targetTile.getTrackConfig();
        // create rotation of target, unless default (= 0) rotation
        if (side != HexSide.get(0)) {
            target = TrackConfig.createByRotation(target, side);
        }
        // check if there are stations to map
        Map<Station, Station> stationMapping = assignStations(base, target);
        if (stationMapping != null && !stationMapping.isEmpty()) {
            if (stationMapping.containsValue(null)) {
                base = TrackConfig.createByDowngrade(base, base.getTile().getStation(1));
            }
            base = TrackConfig.createByStationMapping(base, stationMapping);
        }

        // and finally check if all tracks are maintained
        Set<Track> baseTracks = base.getTracks();
        Set<Track> targetTracks = target.getTracks();
        SetView<Track> diffTrack = Sets.difference(baseTracks, targetTracks);
        if (diffTrack.isEmpty()) {
            SetView<Track> newTracks = Sets.difference(targetTracks, baseTracks);
            boolean allowed = (targetTile.getPossibleRotations().get(side));
            Rotation rotObject = new Rotation(targetTracks, newTracks, side, stationMapping, allowed);
            log.trace("New Rotation for {} => {}: \n{}", baseTile, targetTile, rotObject);
            return rotObject;
        } else {
            log.trace("No Rotation found {} => {}, rotation ={}, remaining Tracks = {}", baseTile, targetTile, side, diffTrack);
            return null;
        }
    }

    private Map<Station, Station> assignStations(TrackConfig base, TrackConfig target) {
        int baseNb = base.getTile().getNumStations();
        if (baseNb == 0) return null;

        int targetNb = target.getTile().getNumStations();
        Map<Station, Station> stationMap = Maps.newHashMapWithExpectedSize(baseNb);
        if (baseNb == 1) {
            // only one station in base => baseStation
            Station baseStation = base.getTile().getStation(1);
            if (targetNb == 1) {
                // only one station in target => targetStation
                Station targetStation = target.getTile().getStation(1);
                // default case: 1 => 1 mapping
                stationMap.put(baseStation, targetStation);
            } else if (targetNb == 0) {
                // special case: downgrade in 1856 and there is only one station to consider
                Set<TrackPoint> baseTrack = base.getStationTracks(baseStation);
                for (TrackPoint side : baseTrack) {
                    Set<TrackPoint> targetTrack = target.getSideTracks((HexSide) side);
                    targetTrack.add(side); // connectivity with all other sides
                    SetView<TrackPoint> diffTrack = Sets.difference(baseTrack, targetTrack);
                    if (!diffTrack.isEmpty()) return null;
                }
                stationMap.put(baseStation, null);
            }
        } else { // more than one base station, assign by side connectivity
            List<Station> noTrackBaseStations = Lists.newArrayList();
            TreeSet<Station> targetStations = Sets.newTreeSet(target.getTile().getStations());
            for (Station b : base.getTile().getStations()) {
                Set<TrackPoint> baseTrack = base.getStationTracks(b);
                if (baseTrack.isEmpty()) { // if track is empty, keep track to add target later
                    noTrackBaseStations.add(b);
                } else {
                    for (Station t : target.getTile().getStations()) {
                        Set<TrackPoint> targetTrack = target.getStationTracks(t);
                        if (checkTrackConnectivity(baseTrack, targetTrack)) {
                            stationMap.put(b, t);
                            targetStations.remove(t);
                            break;
                        }
                    }
                }
            }
            // any base Stations remaining
            for (Station b : noTrackBaseStations) {
                Station t = targetStations.pollFirst();
                if (t != null) {
                    stationMap.put(b, t);
                }
            }
            // check if all base and target stations are assigned
            if (stationMap.keySet().size() != baseNb ||
                    Sets.newHashSet(stationMap.values()).size() != targetNb) {
                stationMap = null;
                log.debug("Mapping: Not all stations assigned, set stationMap to null");
            }
        }
        return stationMap;
    }

    private boolean checkTrackConnectivity(Set<TrackPoint> baseTrack, Set<TrackPoint> targetTrack) {
        SetView<TrackPoint> diffTrack = Sets.difference(baseTrack, targetTrack);
        if (diffTrack.isEmpty()) {
            // target maintains connectivity
            return true;
        } else {
            // if not all connections are maintained,
            Predicate<TrackPoint> checkForStation = new Predicate<TrackPoint>() {
                public boolean apply(TrackPoint p) {
                    return (p.getTrackPointType() == TrackPoint.Type.SIDE);
                }
            };
            // check if remaining tracks only lead to other stations
            if (Sets.filter(diffTrack, checkForStation).isEmpty()) {
                return true;
            }
        }
        return false;
    }


    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("base", baseTile)
                .add("targetTile", targetTile)
                .toString();
    }
}
